### ISSUE TITLE
about the shape loss which is a part of structure loss.

### ISSUE THREAD

**Issue by fornote** [cfzd/Ultra-Fast-Lane-Detection#40]
Hello, thanks for your outstanding works.
I have some problems about the shape loss.
Firstly,while the code calculates the shape loss, it just use the first `dim` instead of all `dim+1` as described in the paper. Therefore, I can't understand your [explanation](https://github.com/cfzd/Ultra-Fast-Lane-Detection/issues/8#issuecomment-647004142) why shape loss would increase in some situations. Because I think the last column isn't used to calculate the expectation of predictions as an approximation of location.

Next, another problem is why the shape loss calculating just uses the first `num_row//2` rows instead of all rows.
This part code in `utils/loss.py` from line 66 to line 71, and I extract these code which is as follws:
```
 for i in range(0,num_rows // 2):
       diff_list1.append(pos[:,i,:] - pos[:,i+1,:])

loss = 0
for i in range(len(diff_list1)-1):
      loss += self.l1(diff_list1[i],diff_list1[i+1])
```

**Comment by cfzd** [comment]
@fornote 
Hi, thanks for your question. 

We use the first `n` (i.e. the first `dim`) elements to locate the lane and the last one (i.e. `dim+1`) element to justify the "no lane" case.  

In this way, for a row that is in "no lane" case, all the output of first `n` elements should be as small as possible, while the last element (i.e. `dim+1`) should be as large as possible. 

The optimization of classification loss can only guarantee that the first `n` elements should be small. However, ***small*** predictions for the first `n` elements ***never mean the correct location***. The location of the lane, which is determined by the first `n` elements, is not guaranteed. 

In other words, the optimization of classification can only make the network output "no lane" (i.e. the `dim+1` element ), but the output of location is not optimized at all. So when we calculate the location of lanes in "no lane"
cases, the results can be just regarded as ***random locations***. So the shape loss might increase.

As for the problem of `num_row//2`, it is a trick which tries to alleviate the above "no lane" problem. Because the first `num_row//2` rows are the output of the results in the ***lower part*** of the image (yes, first row of output corresponds to the last row anchor of image), and the "no lane" case is relatively rare in the lower part of the image, this might reduce the number of "no lane" case and benefit the learning of shape loss. It is the same as only optimize shape loss in the lower part of the image.

**Comment by fornote** [comment]
Thank you very much, I just got it.

**Comment by fornote** [comment]
hello, I just find that I cannot understand what you said that "Because the first `num_row//2` rows are the output of the results in the _**lower part**_ of the image". Why the first `num_row//2` rows are the output of the results in the _**lower part**_ of the image?

Because I found that the **first row** of **class label** whose size was (56,4) corresponded to the **first row** anchor of image. Therefore I have this question.

**Comment by zhaoxin111** [comment]
> hello, I just find that I cannot understand what you said that "Because the first `num_row//2` rows are the output of the results in the _**lower part**_ of the image". Why the first `num_row//2` rows are the output of the results in the _**lower part**_ of the image?
> 
> Because I found that the **first row** of **class label** whose size was (56,4) corresponded to the **first row** anchor of image. Therefore I have this question.

I have carefully check the pipeline of cls_label generation,  the first num_row//2 should be the highest part of the image. Is there any conflict between our understanding and the author’s understanding? @cfzd
